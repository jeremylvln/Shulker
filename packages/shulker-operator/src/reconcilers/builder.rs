use std::fmt::Debug;
use tracing::*;

use kube::{
    api::{DeleteParams, Patch, PatchParams, PostParams},
    Api, ResourceExt,
};
use serde::{de::DeserializeOwned, Serialize};

#[async_trait::async_trait]
pub trait ResourceBuilder {
    type OwnerType: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug;
    type ResourceType: kube::Resource<DynamicType = ()>
        + Clone
        + Serialize
        + DeserializeOwned
        + Debug;

    fn name(owner: &Self::OwnerType) -> String;
    fn api(&self, owner: &Self::OwnerType) -> Api<Self::ResourceType>;

    fn is_needed(&self, _owner: &Self::OwnerType) -> bool {
        true
    }

    fn is_recreation_needed(
        _owner: &Self::OwnerType,
        _existing_resource: &Self::ResourceType,
    ) -> bool {
        false
    }

    async fn build(
        &self,
        owner: &Self::OwnerType,
        name: &str,
        existing_resource: Option<&Self::ResourceType>,
    ) -> Result<Self::ResourceType, anyhow::Error>;
}

async fn get_existing<
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    api: &Api<R>,
    name: &str,
) -> Result<Option<R>, anyhow::Error> {
    api.get(name).await.map(|r| Some(r)).or_else(|e| match e {
        kube::Error::Api(e) => match e.code {
            404 => Ok(None),
            _ => Err(e.into()),
        },
        e => Err(e.into()),
    })
}

fn set_controller_reference<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    owner: &O,
    resource: &mut R,
) {
    resource
        .owner_references_mut()
        .push(owner.controller_owner_ref(&()).unwrap());
}

pub async fn reconcile_builder<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    RB: ResourceBuilder<OwnerType = O, ResourceType = R>,
>(
    builder: &RB,
    owner: &O,
) -> super::Result<Option<R>> {
    let api = builder.api(owner);
    let name = RB::name(owner);

    debug!(
        builder = std::any::type_name::<RB>(),
        owner = std::any::type_name::<O>(),
        owner_name = owner.name_any(),
        target = std::any::type_name::<R>(),
        target_name = name,
        "reconciling builder",
    );

    let mut existing_resource = get_existing(&api, &name)
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    if existing_resource.is_some() {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "found existing resource",
        );

        let delete_existing = || async {
            api.delete(&name, &DeleteParams::default())
                .await
                .map_err(|e| {
                    super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into())
                })
        };

        if !builder.is_needed(owner) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource is not needed anymore, deleting",
            );

            delete_existing().await?;
            return Ok(None);
        } else if RB::is_recreation_needed(owner, existing_resource.as_ref().unwrap()) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource needs to be recreated, deleting",
            );

            delete_existing().await?;
            existing_resource = None;
        }
    }

    if !builder.is_needed(owner) {
        return Ok(None);
    }

    let mut new_resource = builder
        .build(owner, &name, existing_resource.as_ref())
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    let updated_resource = if let Some(existing_resource) = existing_resource {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "patching existing resource",
        );

        for (key, value) in existing_resource.annotations().iter() {
            if !new_resource.annotations().contains_key(key) {
                new_resource
                    .annotations_mut()
                    .insert(key.clone(), value.clone());
            }
        }

        api.patch(
            &name,
            &PatchParams::apply("shulker-operator").force(),
            &Patch::Apply(&new_resource),
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    } else {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "creating new resource",
        );

        set_controller_reference(owner, &mut new_resource);
        api.create(
            &PostParams {
                dry_run: false,
                field_manager: Some("shulker-operator".to_string()),
            },
            &new_resource,
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    };

    Ok(Some(updated_resource))
}
