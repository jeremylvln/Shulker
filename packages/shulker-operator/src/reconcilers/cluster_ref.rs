use kube::{Api, Client};
use shulker_crds::v1alpha1::minecraft_cluster::{MinecraftCluster, MinecraftClusterRef};

use super::{ReconcilerError, Result};

pub async fn resolve_cluster_ref(
    client: &Client,
    namespace: &str,
    cluster_ref: &MinecraftClusterRef,
) -> Result<MinecraftCluster> {
    let clusters_api: Api<MinecraftCluster> = Api::namespaced(client.clone(), namespace);
    let cluster = clusters_api
        .get(&cluster_ref.name)
        .await
        .map_err(|e| ReconcilerError::InvalidClusterRef(cluster_ref.name.clone(), e))?;

    Ok(cluster)
}
