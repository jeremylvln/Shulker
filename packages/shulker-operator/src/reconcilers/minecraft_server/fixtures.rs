use std::collections::BTreeMap;

use http::{Request, Response};
use k8s_openapi::api::core::v1::EnvVar;
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServer, MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_SERVER: MinecraftServer = MinecraftServer {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            tags: vec!["lobby".to_string()],
            version: MinecraftServerVersionSpec {
                channel: MinecraftServerVersion::Paper,
                name: "1.20.1".to_string(),
                ..MinecraftServerVersionSpec::default()
            },
            config: MinecraftServerConfigurationSpec {
                existing_config_map_name: None,
                world: Some(ResourceRefSpec {
                    url: Some("https://example.com/my_world.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }),
                plugins: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_plugin.jar".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                patches: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_patch.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                max_players: 42,
                disable_nether: false,
                disable_end: true,
                server_properties: None,
                proxy_forwarding_mode: MinecraftServerConfigurationProxyForwardingMode::Velocity,
                lifecycle_strategy:
                    MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                skip_agent_download: false
            },
            pod_overrides: Some(MinecraftServerPodOverridesSpec {
                image: None,
                env: Some(vec![EnvVar {
                    name: "EXTRA_ENV".to_string(),
                    value: Some("my_value".to_string()),
                    ..EnvVar::default()
                }]),
                resources: None,
                affinity: None,
                node_selector: Some(BTreeMap::from([(
                    "beta.kubernetes.io/os".to_string(),
                    "linux".to_string()
                )])),
                tolerations: None,
                service_account_name: None,
                volume_mounts: None,
                volumes: None,
                ports: None,
                health: None,
            })
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
