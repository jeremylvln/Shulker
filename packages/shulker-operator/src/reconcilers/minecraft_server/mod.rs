use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::game_server::GameServer;
use k8s_openapi::api::core::v1::ConfigMap;
use kube::{
    api::{DeleteParams, ListParams, PatchParams},
    runtime::{
        controller::Action,
        finalizer::{finalizer, Event as Finalizer},
        watcher::Config,
        Controller,
    },
    Api, Client, ResourceExt,
};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::minecraft_server::{MinecraftServer, MinecraftServerStatus},
};

use self::{config_map::ConfigMapBuilder, gameserver::GameServerBuilder};

use super::{builder::reconcile_builder, status::patch_status, ReconcilerError, Result};

pub mod config_map;
pub mod gameserver;

#[cfg(test)]
mod fixtures;

static FINALIZER: &str = "minecraftservers.shulkermc.io";

struct MinecraftServerReconciler {
    client: kube::Client,

    // Builders
    config_map_builder: ConfigMapBuilder,
    gameserver_builder: GameServerBuilder,
}

impl MinecraftServerReconciler {
    async fn reconcile(
        &self,
        api: Api<MinecraftServer>,
        minecraft_server: Arc<MinecraftServer>,
    ) -> Result<Action> {
        reconcile_builder(&self.config_map_builder, minecraft_server.as_ref()).await?;
        let gameserver =
            reconcile_builder(&self.gameserver_builder, minecraft_server.as_ref()).await?;

        if let Some(gameserver) = &gameserver {
            if let Some(gameserver_status) = &gameserver.status {
                if gameserver_status.state == "Shutdown" {
                    api.delete(&minecraft_server.name_any(), &DeleteParams::default())
                        .await
                        .map_err(ReconcilerError::FailedToDeleteStale)?;

                    return Ok(Action::await_change());
                }

                let mut minecraft_server = minecraft_server.as_ref().clone();
                if minecraft_server.status.is_none() {
                    minecraft_server.status = Some(MinecraftServerStatus::default());
                }

                let status = minecraft_server.status.as_mut().unwrap();

                status.address = gameserver_status.address.clone();
                status.port = gameserver_status.ports.first().unwrap().port;

                if gameserver_status.state == "Ready" || gameserver_status.state == "Allocated" {
                    status.set_condition(
                        "Ready".to_string(),
                        ConditionStatus::True,
                        "ReadyOrAllocated".to_string(),
                        "Server is ready and maybe already allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Ready".to_string(),
                        ConditionStatus::False,
                        "Unknown".to_string(),
                        "Server is not ready yet".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator").force(),
                    &minecraft_server,
                )
                .await?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, minecraft_server: Arc<MinecraftServer>) -> Result<Action> {
        info!(
            name = minecraft_server.name_any(),
            namespace = minecraft_server.namespace(),
            "cleaning up MinecraftServer",
        );

        Ok(Action::await_change())
    }

    fn get_common_labels(minecraft_server: &MinecraftServer) -> BTreeMap<String, String> {
        let fleet_owner = minecraft_server
            .owner_references()
            .iter()
            .find(|owner| owner.kind == "MinecraftServerFleet");

        let mut labels = BTreeMap::from([
            (
                "app.kubernetes.io/name".to_string(),
                minecraft_server.name_any(),
            ),
            (
                "app.kubernetes.io/component".to_string(),
                "minecraftserver".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                minecraft_server.spec.cluster_ref.name.clone(),
            ),
        ]);

        if let Some(owner) = &fleet_owner {
            labels.insert("app.kubernetes.io/name".to_string(), owner.name.clone());
            labels.insert(
                "app.kubernetes.io/instance".to_string(),
                minecraft_server.name_any(),
            );
            labels.insert(
                "minecraftcluster.shulkermc.io/name".to_string(),
                owner.name.clone(),
            );
        }

        labels
    }
}

#[instrument(skip(ctx, minecraft_server))]
async fn reconcile(
    minecraft_server: Arc<MinecraftServer>,
    ctx: Arc<MinecraftServerReconciler>,
) -> Result<Action> {
    let ns = minecraft_server.namespace().unwrap();
    let minecraft_servers_api: Api<MinecraftServer> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = minecraft_server.name_any(),
        namespace = ns,
        "reconciling MinecraftServer",
    );

    finalizer(
        &minecraft_servers_api,
        FINALIZER,
        minecraft_server,
        |event| async {
            match event {
                Finalizer::Apply(minecraft_server) => {
                    ctx.reconcile(minecraft_servers_api.clone(), minecraft_server.clone())
                        .await
                }
                Finalizer::Cleanup(minecraft_server) => ctx.cleanup(minecraft_server.clone()).await,
            }
        },
    )
    .await
    .map_err(|e| ReconcilerError::FinalizerError(Box::new(e)))
}

fn error_policy(
    _minecraft_server: Arc<MinecraftServer>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftServerReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&minecraft_server, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let minecraft_servers_api = Api::<MinecraftServer>::all(client.clone());
    if let Err(e) = minecraft_servers_api
        .list(&ListParams::default().limit(1))
        .await
    {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftServerReconciler {
        client: client.clone(),
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        gameserver_builder: GameServerBuilder::new(client.clone()),
    };

    Controller::new(minecraft_servers_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<GameServer>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
