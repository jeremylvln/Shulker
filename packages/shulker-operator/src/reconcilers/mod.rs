use thiserror::Error;

pub mod builder;
mod cluster_ref;
pub mod minecraft_cluster;
pub mod minecraft_server;
pub mod minecraft_server_fleet;
pub mod proxy_fleet;
pub mod status;

#[derive(Error, Debug)]
pub enum ReconcilerError {
    #[error("failed to reconcile resource: {0}")]
    FinalizerError(#[source] Box<kube::runtime::finalizer::Error<ReconcilerError>>),

    #[error("failed to update resource status: {0}")]
    FailedToUpdateStatus(#[source] kube::Error),

    #[error("failed to delete stale resource: {0}")]
    FailedToDeleteStale(#[source] kube::Error),

    #[error("failed to resolve cluster ref: {1}")]
    InvalidClusterRef(String, #[source] kube::Error),

    #[error("builder {0} failed to build resource: {1}")]
    BuilderError(&'static str, #[source] anyhow::Error),
}

pub type Result<T, E = ReconcilerError> = std::result::Result<T, E>;
