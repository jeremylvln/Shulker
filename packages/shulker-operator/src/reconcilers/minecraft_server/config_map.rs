use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

use super::MinecraftServerReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ConfigMapBuilder {
    type OwnerType = MinecraftServer;
    type ResourceType = ConfigMap;

    fn name(minecraft_server: &Self::OwnerType) -> String {
        format!("{}-config", minecraft_server.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, minecraft_server: &Self::OwnerType) -> bool {
        minecraft_server
            .spec
            .config
            .existing_config_map_name
            .is_none()
    }

    async fn create(
        &self,
        minecraft_server: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerReconciler::get_common_labels(minecraft_server)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..ConfigMap::default()
        };

        Ok(config_map)
    }

    async fn update(
        &self,
        minecraft_server: &Self::OwnerType,
        config_map: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        config_map.data = Some(Self::get_data_from_spec(&minecraft_server.spec.config));
        Ok(())
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }

    pub fn get_data_from_spec(spec: &MinecraftServerConfigurationSpec) -> BTreeMap<String, String> {
        BTreeMap::from([
            (
                "init-fs.sh".to_string(),
                include_str!("../../../assets/server-init-fs.sh").to_string(),
            ),
            (
                "server.properties".to_string(),
                vanilla::VanillaProperties::from_spec(spec).to_string(),
            ),
            (
                "bukkit-config.yml".to_string(),
                bukkit::BukkitYml::from_spec(spec).to_string(),
            ),
            (
                "spigot-config.yml".to_string(),
                spigot::SpigotYml::from_spec(spec).to_string(),
            ),
            (
                "paper-global-config.yml".to_string(),
                paper::PaperGlobalYml::from_spec(spec).to_string(),
            ),
        ])
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_server::fixtures::{create_client_mock, TEST_SERVER},
    };

    #[test]
    fn name_contains_server_name() {
        // W
        let name = super::ConfigMapBuilder::name(&TEST_SERVER);

        // T
        assert_eq!(name, "my-server-config");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);

        // W
        let config_map = builder
            .create(&TEST_SERVER, "my-server-config")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let mut config_map = builder
            .create(&TEST_SERVER, "my-server-config")
            .await
            .unwrap();

        // W
        builder.update(&TEST_SERVER, &mut config_map).await.unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[test]
    fn get_data_from_spec_has_startup_script() {
        // G
        let spec = TEST_SERVER.spec.config.clone();

        // W
        let data = super::ConfigMapBuilder::get_data_from_spec(&spec);

        // T
        assert!(data.contains_key("init-fs.sh"));
    }

    #[test]
    fn get_data_from_spec_has_server_configs() {
        // G
        let spec = TEST_SERVER.spec.config.clone();

        // W
        let data = super::ConfigMapBuilder::get_data_from_spec(&spec);

        // T
        assert!(data.contains_key("server.properties"));
        assert!(data.contains_key("bukkit-config.yml"));
        assert!(data.contains_key("spigot-config.yml"));
        assert!(data.contains_key("paper-global-config.yml"));
    }
}

mod vanilla {
    use std::collections::BTreeMap;

    use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

    pub struct VanillaProperties(BTreeMap<String, String>);

    impl VanillaProperties {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            let mut properties = BTreeMap::new();

            properties.insert("max-players".to_string(), spec.max_players.to_string());
            properties.insert(
                "allow-nether".to_string(),
                (!spec.disable_nether).to_string(),
            );
            properties.insert("online-mode".to_string(), "false".to_string());
            properties.insert("prevent-proxy-connections".to_string(), "false".to_string());
            properties.insert("enforce-secure-profiles".to_string(), "true".to_string());

            VanillaProperties(properties)
        }
    }

    impl ToString for VanillaProperties {
        fn to_string(&self) -> String {
            let mut properties = String::new();

            for (key, value) in &self.0 {
                properties.push_str(&format!("{}={}\n", key, value));
            }

            properties
        }
    }

    #[cfg(test)]
    mod tests {
        use std::collections::BTreeMap;

        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::VanillaProperties::from_spec(&spec);

            // T
            insta::assert_debug_snapshot!(config.0);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::VanillaProperties(BTreeMap::from([(
                "my-str".to_string(),
                "my-value".to_string(),
            )]));

            // W
            let properties = config.to_string();

            // T
            insta::assert_snapshot!(properties);
        }
    }
}

mod bukkit {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitYml {
        settings: BukkitSettingsYml,
        auto_updater: BukkitAutoUpdaterYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitSettingsYml {
        allow_end: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitAutoUpdaterYml {
        enabled: bool,
    }

    impl BukkitYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            BukkitYml {
                settings: BukkitSettingsYml {
                    allow_end: !spec.disable_end,
                },
                auto_updater: BukkitAutoUpdaterYml { enabled: false },
            }
        }
    }

    impl ToString for BukkitYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::BukkitYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::BukkitYml {
                settings: super::BukkitSettingsYml { allow_end: true },
                auto_updater: super::BukkitAutoUpdaterYml { enabled: false },
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}

mod spigot {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::{
        MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotYml {
        settings: SpigotSettingsYml,
        advancements: SpigotSaveableYml,
        players: SpigotSaveableYml,
        stats: SpigotSaveableYml,
        save_user_cache_on_stop_only: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotSettingsYml {
        bungeecord: bool,
        restart_on_crash: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotSaveableYml {
        disable_saving: bool,
    }

    impl SpigotYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            SpigotYml {
                settings: SpigotSettingsYml {
                    bungeecord: spec.proxy_forwarding_mode
                        == MinecraftServerConfigurationProxyForwardingMode::BungeeCord,
                    restart_on_crash: false,
                },
                advancements: SpigotSaveableYml {
                    disable_saving: true,
                },
                players: SpigotSaveableYml {
                    disable_saving: true,
                },
                stats: SpigotSaveableYml {
                    disable_saving: true,
                },
                save_user_cache_on_stop_only: true,
            }
        }
    }

    impl ToString for SpigotYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::SpigotYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::SpigotYml {
                settings: super::SpigotSettingsYml {
                    bungeecord: false,
                    restart_on_crash: false,
                },
                advancements: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                players: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                stats: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                save_user_cache_on_stop_only: true,
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}

mod paper {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::{
        MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalYml {
        proxies: PaperGlobalProxiesYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesYml {
        bungee_cord: PaperGlobalProxiesBungeeCordYml,
        velocity: PaperGlobalProxiesVelocityYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesBungeeCordYml {
        online_mode: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesVelocityYml {
        enabled: bool,
        online_mode: bool,
        secret: String,
    }

    impl PaperGlobalYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            PaperGlobalYml {
                proxies: PaperGlobalProxiesYml {
                    bungee_cord: PaperGlobalProxiesBungeeCordYml {
                        online_mode: spec.proxy_forwarding_mode
                            == MinecraftServerConfigurationProxyForwardingMode::BungeeCord,
                    },
                    velocity: PaperGlobalProxiesVelocityYml {
                        enabled: spec.proxy_forwarding_mode
                            == MinecraftServerConfigurationProxyForwardingMode::Velocity,
                        online_mode: true,
                        secret: "${CFG_VELOCITY_FORWARDING_SECRET}".to_string(),
                    },
                },
            }
        }
    }

    impl ToString for PaperGlobalYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::PaperGlobalYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::PaperGlobalYml {
                proxies: super::PaperGlobalProxiesYml {
                    bungee_cord: super::PaperGlobalProxiesBungeeCordYml { online_mode: false },
                    velocity: super::PaperGlobalProxiesVelocityYml {
                        enabled: true,
                        online_mode: true,
                        secret: "my-secret".to_string(),
                    },
                },
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}
