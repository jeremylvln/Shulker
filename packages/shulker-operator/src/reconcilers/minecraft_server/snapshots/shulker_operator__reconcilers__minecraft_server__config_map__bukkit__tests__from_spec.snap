---
source: packages/shulker-operator/src/reconcilers/minecraft_server/config_map.rs
expression: config
---
settings:
  allow-end: false
auto-updater:
  enabled: false

